relayer/src/worker/connection.rs

```diff
@@ -0,0 +1,127 @@
+use std::{thread, time::Duration};
+
+use anomaly::BoxError;
+use crossbeam_channel::Receiver;
+use tracing::{debug, warn};
+
+use crate::connection::Connection as RelayConnection;
+use crate::telemetry::Telemetry;
+use crate::{
+    chain::handle::ChainHandlePair, object::Connection, util::retry::retry_with_index,
+    worker::retry_strategy,
+};
+
+use super::WorkerCmd;
+
+pub struct ConnectionWorker {
+    connection: Connection,
+    chains: ChainHandlePair,
+    cmd_rx: Receiver<WorkerCmd>,
+
+    #[allow(dead_code)]
+    telemetry: Telemetry,
+}
+
+impl ConnectionWorker {
+    pub fn new(
+        connection: Connection,
+        chains: ChainHandlePair,
+        cmd_rx: Receiver<WorkerCmd>,
+        telemetry: Telemetry,
+    ) -> Self {
+        Self {
+            connection,
+            chains,
+            cmd_rx,
+            telemetry,
+        }
+    }
+
+    /// Run the event loop for events associated with a [`Connection`].
+    pub(crate) fn run(self) -> Result<(), BoxError> {
+        let a_chain = self.chains.a.clone();
+        let b_chain = self.chains.b.clone();
+
+        // Flag that indicates if the worker should actively resume handshake.
+        // Set on start or when event based handshake fails.
+        let mut resume_handshake = true;
+
+        loop {
+            thread::sleep(Duration::from_millis(200));
+
+            if let Ok(cmd) = self.cmd_rx.try_recv() {
```

Comment:
If `try_recv()` returns `TryRecvError::Disconnected`, the sender side of the channel has been dropped and the loop is going to keep failing forever.

Reply:
yes this is true. All workers have the same pattern. The supervisor should not disconnect unless it has crashed and in this case not much can be done. But we could indeed check this error and exit. I think we discussed with @romac to have a way for the supervisor to gracefully close the worker, not sure this is done. I think the worker has a way to gracefully exit so the supervisor can clean its worker info.